pkg/scm/git/git.go

```diff
@@ -62,18 +69,145 @@ func stringInSlice(s string, slice []string) bool {
+func (h *stiGit) ValidCloneSpecRemoteOnly(source string) bool {
+	rc, _, _, _, _, _ := validCloneSpec(source, true, nil)
+	return rc
+}
+
+// MungeNoProtocolURL will take a URL returned from net/url.Parse and make
+// corrections to the URL when no protocol is specified in the Scheme, where there
+// are valid protocol-less git url spec formats that result in either file:// or ssh:// protocol usage;
+// if an explicit protocol other than file:// or ssh:// are already specified in a valid way, then the
+// URL is left unchaged and the method simply returns with no error reported,
+// since the URL is
+func (h *stiGit) MungeNoProtocolURL(source string, uri *url.URL) error {
+	if uri == nil {
+		return nil
+	}
+	rc, scheme, user, host, path, ref := validCloneSpec(source, false, uri)
+	if !rc {
+		return errors.New(fmt.Sprintf("invalid git clone spec provided, source string %s and URL %v", source, uri))
+	}
+	// don't modify the protocol options which always have to be explicitly specified (http, https,  git)
+	if scheme != "ssh" && scheme != "file" {
+		return nil
+	}
+
+	// update the either file or ssh url accordingly
+	if len(user) > 0 {
+		uri.User = url.User(user)
+	}
+	uri.Scheme = scheme
+	uri.Host = host
+	uri.Path = path
+	uri.Fragment = ref
+	return nil
+}
+
+func schemeNecessitatesReturn(uri *url.URL) (shouldReturn, returnCode bool) {
+	if uri == nil {
+		return false, true
+	}
+
+	if stringInSlice(uri.Scheme, allowedSchemes) {
+		return true, true
+	}
+
+	//at this point, the only valid value for Scheme should be "";  if it is
+	// set to some other protocol besides what is listed in allowedSchemes,
+	// say ftp://, then we return false, as it is a protocol we are not expecting.
+	// note, there also appears to be a bug in url.Parse, where a url like
+	// linux-box:/path/to/code will report the scheme as "linux-box"
+	if uri.Scheme != "" {
+		return true, false
+	}
+
+	return false, true
+}
+
+func validCloneSpec(source string, remoteOnly bool, partialURL *url.URL) (rc bool, scheme, user, host, path, ref string) {
+	// note, as part of parsing the clone spec to see if it is valid, various values for the fields in url.URL are saved
+	// and returned for possible URL corrections from MungeNoProtocolURL
+	if partialURL == nil {
+		uri, err := url.Parse(source)
+		if err != nil {
+			return false, scheme, user, host, path, ref
+		}
+
+		// ValidCloneRemoteOnly only goes down the partialURL == nil path
+		if uri.Scheme == "file" && remoteOnly {
+			return false, scheme, user, host, path, ref
+		}
+
+		shouldReturn, returnCode := schemeNecessitatesReturn(uri)
+		if shouldReturn {
+			return returnCode, scheme, user, host, path, ref
+		}
+
+	} else {
+		shouldReturn, returnCode := schemeNecessitatesReturn(partialURL)
+		if shouldReturn {
+			return returnCode, scheme, user, host, path, ref
+		}
 	}
 
-	if stringInSlice(url.Scheme, allowedSchemes) {
-		return true
+	lastColonIdx := strings.LastIndex(source, ":")
+	if lastColonIdx != -1 {
```

Comment:
Ah - I see now.  Yep, you are right.

And y'all are probably aware, but just in case, today the s2i code does a golang os level copy of files for either file:// or just plain paths.

So to me it would make sense to allow ":" in both those cases.  Agreed?

The trick will be distinguishing between protocol-less file spec and ssh spec.